lib/core/include/scipp/core/sizes.h

```diff
@@ -51,42 +55,41 @@ class SCIPP_CORE_EXPORT small_stable_map {
   void erase(const Key &key);
   void clear() noexcept;
   void replace_key(const Key &from, const Key &to);
-  [[nodiscard]] constexpr scipp::span<const Key> keys() const &noexcept {
+  [[nodiscard]] scipp::span<const Key> keys() const &noexcept {
     return {m_keys.data(), static_cast<size_t>(size())};
   }
-  [[nodiscard]] constexpr scipp::span<const Value> values() const &noexcept {
+  [[nodiscard]] scipp::span<const Value> values() const &noexcept {
     return {m_values.data(), static_cast<size_t>(size())};
   }
 
 private:
-  int16_t m_size{0};
-  std::array<Key, Capacity> m_keys{};
-  std::array<Value, Capacity> m_values{};
+  boost::container::small_vector<Key, Capacity> m_keys{};
+  boost::container::small_vector<Value, Capacity> m_values{};
```

Comment:
Did you consider using `small_vector<std::pair<Key, Value>, Capacity>` in order to keep the data close together on the heap?

Reply:
That would complicate returning iterators to keys and values (would need a custom iterator, probably `boost::iterator::transform_iterator` would do).

I don't think keeping things closer on the heap is the main reason. But there are two others: a) 1 instead of 2 heap allocations (or 1 instead of 3, if we integrate strides). b) Less size overhead on the stack, by having only 1 instead of 2 `small_vector` instances (but this is traded against losing some space to padding).

I think we can consider this for later, in combination with moving stride handling into `Dimensions`.